src/vm/collection.rs

```diff
@@ -112,28 +112,43 @@ pub trait Collection<VM: VMBinding> {
     /// * `tls_worker`: The thread pointer for the worker thread performing this call.
     fn post_forwarding(_tls: VMWorkerThread) {}
 
-    /// Return the amount of memory (in pages) the VM allocated off-heap which could be released
-    /// by doing GC.  MMTk core will trigger a GC if the amount of allocated memory in MMTk spaces
-    /// plus this amount exceeds a certain limit.
+    /// Return the amount of memory (in bytes) the VM allocated (but not released yet) outside the
+    /// MMTk heap which can be released by doing GC.
+    ///
+    /// MMTk core will add this amount to the amount of memory allocated or reserved in MMTk
+    /// spaces, and will trigger a GC if the sum exceeds a certain limit.
```

Comment:
How to trigger a GC depends on the internal GC trigger code. You could simply say that MMTk core will consider the reported memory as part of MMTk heap for the purpose of heap size accounting.

Reply:
Yes. It is the GC trigger that determines when to trigger GC, but I want the VM binding developer to know that the return value of this function is additive (rather than multiplicative or having other non-linear relation) to MMTk core's "used pages" when considered to trigger GC, so that the developer can have a basic idea of what effect the return value of this function will have on triggering GC.  At least the current implementation is additive for both fixed heap size and the MemBalancer.  The MemBalancer changes the heap limit dynamically, but the way to compute occupied pages is still adding the used pages of MMTk spaces and the result of this function.  I can emphasise that being additive is only a detail of the current implementation.

Reply:
> I want the VM binding developer to know that the return value of this function is additive (rather than multiplicative or having other non-linear relation) to MMTk core's "used pages" when considered to trigger GC

I think this is incorrect. GC triggering does not necessarily relate to heap sizes. We could implement a GC trigger that uses allocation volume -- in which case, we do not directly use the vm allocated pages returned by the method, we may use a diff of value or simply ignore it. Sticky immix does not trigger nursery GC based on heap sizes, LXR has similar behaviors as well.
https://github.com/mmtk/mmtk-core/blob/40c977308cb24d8bdf46b82916a388c10fa956e7/src/plan/sticky/immix/global.rs#L153

This is what I suggest about the contract of this method.
> MMTk core will consider the reported memory as part of MMTk heap for the purpose of heap size accounting.

But we do not guarantee that GC triggering is related with heap size.

Reply:
I removed the second paragraph and actually give recommendations for VMs about what to include instead of saying "MMTk doesn't specify what to include except that number is considered by the GC trigger".  In this way, newer GC triggers can decide whether to take that reported number into consideration when triggering GC.